### PR TITLE
[Validator] fixed duplicate constraints with parent class interfaces

### PR DESCRIPTION
**Pull request by dmaicher**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/19516 |
| License | MIT |
| Doc PR | - |

This fixes https://github.com/symfony/symfony/issues/19516

### CODE REVIEW

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php [hunk 1]
```diff
@@ -116,9 +116,12 @@ public function getMetadataFor($value)
             $metadata->mergeConstraints($this->getMetadataFor($parent->name));
         }
 
-        // Include constraints from all implemented interfaces
+        // Include constraints from all implemented interfaces that have not been processed via parent class yet
         foreach ($metadata->getReflectionClass()->getInterfaces() as $interface) {
-            if ('Symfony\Component\Validator\GroupSequenceProviderInterface' === $interface->name) {
+            if (
+                'Symfony\Component\Validator\GroupSequenceProviderInterface' === $interface->name ||
+                $parent && $parent->implementsInterface($interface->name)
```
Comment (HeahDude) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: Does it work for a child of a child?

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php [hunk 2]
```diff
@@ -116,9 +116,12 @@ public function getMetadataFor($value)
...
+            ) {
```
Comment (iltar) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: Not sure if the codestyle of the if is 100% correct, though fabbot.io is not complaining

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php (hunk 1 above)
Comment (dmaicher) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: I think it should work :wink: No matter at which level (starting at the parent then going upwards to the parent of the parent etc.) the interface is implemented this condition will be true.

Or what exactly do you mean?

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php (hunk 2 above)
Comment (dmaicher) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: Was also not 100% sure about it :blush: Any preference how to make it look nicer?

Comment (iltar) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: I think it was either everything on 1 line or 

```php
if ($a
    || $b
    || $c
) {
```
But I think it's simply 1 line for Symfony

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php (hunk 1 above)
Comment (HeahDude) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: I mean I was wondering about the pertinence of adding a test for it but you look pretty sure :)

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: 👍 on multiline

Comment (dmaicher) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: Changed it :+1:

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php [hunk 3]
```diff
@@ -116,9 +116,11 @@ public function getMetadataFor($value)
             $metadata->mergeConstraints($this->getMetadataFor($parent->name));
         }
 
-        // Include constraints from all implemented interfaces
+        // Include constraints from all implemented interfaces that have not been processed via parent class yet
         foreach ($metadata->getReflectionClass()->getInterfaces() as $interface) {
-            if ('Symfony\Component\Validator\GroupSequenceProviderInterface' === $interface->name) {
+            if ('Symfony\Component\Validator\GroupSequenceProviderInterface' === $interface->name
+                || $parent && $parent->implementsInterface($interface->name)
```
Comment (jakzal) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: I'd keep the condition in a single line, but add parenthesis to make precedence explicit.

**Comment by dmaicher** [comment]
@webmozart you are probably the most qualified to say if this fix is correct? :blush:

File: src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php (hunk 3 above)
Comment (fabpot) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: We don't enforce line length in Symfony, so keeping everything on one line is the way to go.

Comment (dmaicher) [inline src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php]: I changed it :+1:

**Comment by dmaicher** [comment]
@fabpot @stof @nicolas-grekas who could have a look at this? :blush:

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @dmaicher.

**Pull request merged**